src/types/ErrorRate.ts

```diff
@@ -0,0 +1,335 @@
+        if (auxStatus.status.priority > result.status.status.priority) {
+          result.status = auxStatus;
+          result.protocol = protocol;
+          result.toleranceConfig = reqTol.tolerance;
+        }
+      }
+    }
+  }
+  return result;
+};
+
+const generateRateForTolerance = (tol: RequestTolerance, requests: { [key: string]: { [key: string]: number } }) => {
+  for (let [protocol, req] of Object.entries(requests)) {
+    if (tol.tolerance) {
+      if (checkExpr(tol!.tolerance!.protocol, protocol)) {
+        for (let [code, value] of Object.entries(req)) {
+          if (!Object.keys(tol.requests).includes(protocol)) {
+            tol.requests[protocol] = emptyRate();
+          }
+          tol.requests[protocol].requestRate += Number(value);
+          if (checkExpr(tol!.tolerance!.code, code)) {
+            tol.requests[protocol].errorRate += Number(value);
+          }
+        }
+      }
+    } else {
+      for (let [code, value] of Object.entries(req)) {
+        if (!Object.keys(tol.requests).includes(protocol)) {
+          tol.requests[protocol] = emptyRate();
+        }
+        tol.requests[protocol].requestRate += Number(value);
+        if (Object.keys(regexProtocol).includes(protocol) && regexProtocol[protocol].test(code)) {
+          tol.requests[protocol].errorRate += Number(value);
+        }
+      }
+    }
+    if (Object.keys(tol.requests).includes(protocol)) {
+      if (tol.requests[protocol].requestRate === 0) {
+        tol.requests[protocol].errorRatio = -1;
+      } else {
+        tol.requests[protocol].errorRatio = tol.requests[protocol].errorRate / tol.requests[protocol].requestRate;
+      }
+    }
+  }
+};
+
+const aggregate = (request: RequestType, tolerances?: ToleranceConfig[]): RequestTolerance[] => {
+  let result: RequestTolerance[] = [];
+  if (request) {
+    if (tolerances) {
+      for (let tol of Object.values(tolerances)) {
+        let newReqTol: RequestTolerance = { tolerance: tol, requests: {} };
+        generateRateForTolerance(newReqTol, request);
+        result.push(newReqTol);
+      }
+    } else {
+      const emptyRequest: { [key: string]: Rate } = {};
+      for (let protocol of Object.keys(request)) {
+        emptyRequest[protocol] = emptyRate();
+      }
+      let newReqTol: RequestTolerance = { tolerance: undefined, requests: emptyRequest };
+      generateRateForTolerance(newReqTol, request);
+      result.push(newReqTol);
+    }
+  }
+  return result;
+};
+
+const emptyRate = (): Rate => {
+  return { requestRate: 0, errorRate: 0, errorRatio: 0 };
+};
+
+// Graph Edge
+
+export const graphEdgeColor = (edge: DecoratedGraphEdgeData): ThresholdStatus => {
+  if (edge && edge.responses) {
+    const cy = store.getState().graph.summaryData!.summaryTarget as Cy.Core;
+    const source = decoratedNodeData(cy.$id(edge.source));
+    const target = decoratedNodeData(cy.$id(edge.target));
```

Comment:
These lines should not be needed if you take my suggestion about the function signature change.